lib/rules/async-server-action.js

```diff
@@ -0,0 +1,57 @@
+/**
+ * @fileoverview Require functions with the `use server` directive to be async
+ * @author Jorge Zreik
+ */
+
+'use strict';
+
+const docsUrl = require('../util/docsUrl');
+const report = require('../util/report');
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+const messages = {
+  asyncServerAction: 'Server Actions must be async',
+  suggestAsync: 'Make {{functionName}} async',
+};
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Require functions with the `use server` directive to be async',
+      category: 'Possible Errors',
+      recommended: false,
+      url: docsUrl('async-server-action'),
+    },
+
+    messages,
+
+    type: 'suggestion',
+    hasSuggestions: true,
+
+    schema: [],
+  },
+
+  create(context) {
+    return {
+      ':function[async=false]>BlockStatement>:first-child[expression.value="use server"]'(node) {
+        const currentFunction = node.parent.parent;
+        const functionName = currentFunction.id ? `\`${currentFunction.id.name}\`` : 'this function';
+        report(context, messages.asyncServerAction, 'asyncServerAction', {
+          node: currentFunction,
+          data: {
+            functionName,
+          },
+          suggest: [{
+            desc: messages.suggestAsync,
+            fix(fixer) {
+              return fixer.insertTextBefore(currentFunction, 'async ');
```

Comment:
can we also ensure that this won't apply to generators?

Reply:
Is there a reason this can't be the case? I'd imagine that if someone added the `"use server"` directive to a generator function it should still be async.

Reply:
can components even *be* generator functions?

also, async generators vs generators is a very different semantic shift than async functions vs functions.

Reply:
Changed!